### PR TITLE
Removing the magic invisibility of shady zombies

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Balance "Removes the magic invisibility of shady zombies"```

#### Purpose of change
Magic abilities in the base game are kept to a minimum. However, shady zombies only become visible one tile away. There is no way to notice a shady zombie before it reaches the player.

The shady zombie's night invisibility is best described as magic, rather than camouflage.

It is strange for shady zombies to be visible at the same range for characters with 20+ perception and characters with 4.

#### Describe the solution
This PR changes the range at which shady zombies are noticed. The magic-invisibility has been replaced by a perception-dependent camouflage.

As per the original PR (#14322), strong light sources still break the shady zombie's camouflage. AFAICT the original author's intention was to give the shady zombie some kind of camouflage/light-bending ability which prevents night vision from seeing it. 

Perception is now the way to detect a shady zombie; instead of trying to see past the light-bending ability, high perception characters notice the distortions/abnormalities surrounding the shady zombie and discover it's position that way.

If a shady zombie is more than ( (perception - (eye encumbrance / 5) ) / 3 ) tiles away (both divisions are rounded down), the shady zombie will be invisible.

A side effect of this PR will increase the usefulness of perception, which is arguably one of the least useful stats.

A min-maxed character with 4 perception will only see shady zombies from 1 tile away
An average character with 8 perception will notice it from 2.
A character who adds 1 point to perception at chargen will see it from 3 away
And characters with 12 (the highest amount that most players will choose to begin with) will see shady zombies from 4 tiles away.

#### Describe alternatives you've considered
Raising the required perception per tile to 4

#### Additional context
Pictures:
Note: the shady zombies are in a line extending to the right of the player

image
4 perception (minmaxer)

image
8 perception (average)

image
9 perception (above average)

image
12 perception (ranged master)

image
20 perception (maxed base perception)

If repository maintainers want to edit this, they can.

### CODE REVIEW

**Comment by alkemann** [comment]
Was going to play test, but does not compile

**Comment by SourSlime** [comment]
Did you get some kind of error? 

I’d compile it if I could, but I have no idea how to do it on a mac

**Comment by anothersimulacrum** [comment]
It's actually [pretty easy on a Mac](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/COMPILING.md#mac-os-x).

**Comment by alkemann** [comment]
@SourSlime You can see the error also on the [Travis build](https://travis-ci.org/CleverRaven/Cataclysm-DDA/jobs/576544280):
```
src/creature.cpp:179:16: error: out-of-line definition of 'sees' does not match any declaration in 'Creature'
bool Creature::sees( const Creature &critter, int perception ) const
```

File: src/creature.cpp [hunk 1]
```diff
@@ -207,8 +207,9 @@ bool Creature::sees( const Creature &critter ) const
     if( wanted_range <= 1 &&
         ( posz() == critter.posz() || g->m.valid_move( pos(), critter.pos(), false, true ) ) ) {
         return visible( p );
+    } else if( ( wanted_range > (perception / 3) && critter.has_flag( MF_NIGHT_INVISIBILITY ) && g->m.light_at( critter.pos() ) <= LL_LOW ) ) {
```
Comment (LaVeyanFiend) [inline src/creature.cpp]: Does that treat NV towards the perception number?

Comment (anothersimulacrum) [reply src/creature.cpp]: No.

Comment (SourSlime) [reply src/creature.cpp]: No, but see the updated original post. I believe that shady zombies have a special, blob-given ability that passively distorts light in shady areas, rendering the zombie invisible (think of this ability similar to the passive effect of an incandesant husk).

High-perception characters notice the distortion around the shady zombie, which allows them to locate where it is, despite not really seeing it (they would see, or perhaps imagine, a vague outline of the body).

**Comment by SourSlime** [comment]
Thanks for all the help guys, I'm pretty sure I got it to work. I managed to compile Cata successfully on my mac, but I have no idea what happened to the app.

File: src/creature.cpp [hunk 2]
```diff
@@ -207,8 +207,10 @@ bool Creature::sees( const Creature &critter ) const
     if( wanted_range <= 1 &&
         ( posz() == critter.posz() || g->m.valid_move( pos(), critter.pos(), false, true ) ) ) {
         return visible( p );
+    } else if( ( wanted_range > ( ( p->get_per() - ( p->encumb( bp_eyes ) / 5.0f ) ) / 3.0f ) ) && 
+               ( critter.has_flag( MF_NIGHT_INVISIBILITY ) ) && g->m.light_at( critter.pos() ) <= LL_LOW ) {
```
Comment (BevapDin) [inline src/creature.cpp]:
```suggestion
               critter.has_flag( MF_NIGHT_INVISIBILITY ) && g->m.light_at( critter.pos() ) <= LL_LOW ) {
```

File: src/creature.cpp [hunk 3]
```diff
@@ -207,8 +207,10 @@ bool Creature::sees( const Creature &critter ) const
     if( wanted_range <= 1 &&
         ( posz() == critter.posz() || g->m.valid_move( pos(), critter.pos(), false, true ) ) ) {
         return visible( p );
+    } else if( ( wanted_range > ( ( p->get_per() - ( p->encumb( bp_eyes ) / 5.0f ) ) / 3.0f ) ) &&
```
Comment (BevapDin) [inline src/creature.cpp]: `p` can be null. You must check for this before actually using it. And `p` is the creature that is looked at (the functions checks whether `p` / `critter` can be seen). The visibility of someone should not depend on their perception. (Note: `this` is the creature that is actually looking.)

**Comment by SourSlime** [comment]
Thanks everyone for the help.

I learnt a fair bit of programming this week, and I managed to get my build working properly, so I was able to test the PR

I think everything should be working now. It turns out I was adding stuff to wrong the file, so I reverted that and added it to the right one.

Since it's working, I might be able to make a gif showing the changes

File: src/creature.cpp (hunk 2 above)
Comment (SourSlime) [reply src/creature.cpp]: Thanks

File: src/creature.cpp (hunk 3 above)
Comment (SourSlime) [reply src/creature.cpp]: Yeah, I don't really know how to program lol.

Thanks for the help though, it should be working now :)

File: src/player.cpp [hunk 4]
```diff
@@ -11197,11 +11197,15 @@ bool player::sees( const tripoint &t, bool, int ) const
 
 bool player::sees( const Creature &critter ) const
 {
-    // This handles only the player/npc specific stuff (monsters don't have traits or bionics).
+    // This handles only the player/npc specific stuff (monsters don't have traits, stats or bionics).
     const int dist = rl_dist( pos(), critter.pos() );
     if( dist <= 3 && has_active_mutation( trait_ANTENNAE ) ) {
         return true;
     }
+    if( dist <= ( ( get_per() - ( encumb( bp_eyes ) / 5.0f ) ) / 3.0f ) &&
```
Comment (BevapDin) [inline src/player.cpp]:
```suggestion
    if( dist <= ( get_per() - encumb( bp_eyes ) / 5.0f ) / 3 &&
```
Or:
```C++
    const float actual_per = get_per() - encumb( bp_eyes ) / 5.0f;
    if( dist <= actual_per / 3 &&
```

Comment (SourSlime) [reply src/player.cpp]: Thanks for the help

**Comment by SourSlime** [comment]
Yes, I suspected that as well, but did I run some tests before you commented, and I wasn't able to see shady zombies through walls, so I thought it wouldn't be a problem.

I decided to run some more extensive tests though, and I found some things:
-Shady Zombies can be seen through walls like you said, but they only show up on the V-menu  or the compass. The sprite doesn't appear (which is why I thought it was working properly before), and looking at the tile with x shows "Unseen". Using the V-menu can pinpoint the monster's location however.
-The same thing seems to happen with Terranian Sonar and Infrared Vision, which are supposed to work through walls

I'll open a issue about the Terranian Sonar stuff.

It should probably work now though, since the player has to be able to see the tile to see the shady zombie. I only gave a quick test, but the shady zombie doesn't appear in the compass/V-menu anymore.

**Pull request closed without merging**